# **Movie Dataset Python Assignment**


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### **Question 1**
Dataset (`movie.csv`) ko pandas mein load karein aur uske pehle 5 rows display karein.

In [4]:
df = pd.read_csv('movie.csv')
df.head()

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
0,The Shawshank Redemption,1994,A,142 min,Drama,9.3,80.0,Frank Darabont,Tim Robbins,2343110,28341469.0
1,The Godfather,1972,A,175 min,"Crime, Drama",9.2,100.0,Francis Ford Coppola,Marlon Brando,1620367,134966411.0
2,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,84.0,Christopher Nolan,Christian Bale,2303232,534858444.0
3,The Godfather: Part II,1974,A,202 min,"Crime, Drama",9.0,90.0,Francis Ford Coppola,Al Pacino,1129952,57300000.0
4,12 Angry Men,1957,U,96 min,"Crime, Drama",9.0,96.0,Sidney Lumet,Henry Fonda,689845,4360000.0


### **Question 2**
Dataset ka total shape (rows aur columns count) check karke print karein.

In [6]:
df.shape  

(1000, 11)

### **Question 3**
Check karein ki kis kis column mein missing values (null values) hain aur unka exact count kya hai.

In [23]:
df.isnull().sum()

Series_Title       0
Released_Year      0
Certificate      101
Runtime            0
Genre              0
IMDB_Rating        0
Meta_score       157
Director           0
Star1              0
No_of_Votes        0
Gross            169
dtype: int64

### **Question 4**
Which movies were released after the year 2000? (Year 2000 ke baad release hone wali movies filter karein).

In [25]:
df[df['Released_Year'] > "2000"]

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
2,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,84.0,Christopher Nolan,Christian Bale,2303232,534858444.0
5,The Lord of the Rings: The Return of the King,2003,U,201 min,"Action, Adventure, Drama",8.9,94.0,Peter Jackson,Elijah Wood,1642758,377845905.0
8,Inception,2010,UA,148 min,"Action, Adventure, Sci-Fi",8.8,74.0,Christopher Nolan,Leonardo DiCaprio,2067042,292576195.0
10,The Lord of the Rings: The Fellowship of the Ring,2001,U,178 min,"Action, Adventure, Drama",8.8,92.0,Peter Jackson,Elijah Wood,1661481,315544750.0
13,The Lord of the Rings: The Two Towers,2002,UA,179 min,"Action, Adventure, Drama",8.7,87.0,Peter Jackson,Elijah Wood,1485555,342551365.0
...,...,...,...,...,...,...,...,...,...,...,...
948,The Others,2001,PG-13,101 min,"Horror, Mystery, Thriller",7.6,74.0,Alejandro Amenábar,Nicole Kidman,337651,96522687.0
949,Blow,2001,R,124 min,"Biography, Crime, Drama",7.6,52.0,Ted Demme,Johnny Depp,240714,52990775.0
950,Enemy at the Gates,2001,A,131 min,"Drama, History, War",7.6,53.0,Jean-Jacques Annaud,Jude Law,243729,51401758.0
951,Minority Report,2002,A,145 min,"Action, Crime, Mystery",7.6,80.0,Steven Spielberg,Tom Cruise,508417,132072926.0


### **Question 5**
Which movies have an IMDB rating greater than 9? (9 se zyada IMDB rating wali movies display karein).

In [27]:
df[df['IMDB_Rating'] > 9]

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
0,The Shawshank Redemption,1994,A,142 min,Drama,9.3,80.0,Frank Darabont,Tim Robbins,2343110,28341469.0
1,The Godfather,1972,A,175 min,"Crime, Drama",9.2,100.0,Francis Ford Coppola,Marlon Brando,1620367,134966411.0


### **Question 6**
How many movies are there for each genre? (Har genre ki total kitni movies hain, count nikalye).

In [31]:
df['Genre'].value_counts()

Genre
Drama                        85
Drama, Romance               37
Comedy, Drama                35
Comedy, Drama, Romance       31
Action, Crime, Drama         30
                             ..
Action, Adventure, Family     1
Action, Crime, Mystery        1
Animation, Drama, Romance     1
Drama, War, Western           1
Adventure, Comedy, War        1
Name: count, Length: 202, dtype: int64

### **Question 7**
Which movies were directed by Christopher Nolan? (Christopher Nolan ki directed saari movies nikalye).

In [32]:
df[df['Director'] == "Christopher Nolan"]

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
2,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,84.0,Christopher Nolan,Christian Bale,2303232,534858444.0
8,Inception,2010,UA,148 min,"Action, Adventure, Sci-Fi",8.8,74.0,Christopher Nolan,Leonardo DiCaprio,2067042,292576195.0
21,Interstellar,2014,UA,169 min,"Adventure, Drama, Sci-Fi",8.6,74.0,Christopher Nolan,Matthew McConaughey,1512360,188020017.0
36,The Prestige,2006,U,130 min,"Drama, Mystery, Sci-Fi",8.5,66.0,Christopher Nolan,Christian Bale,1190259,53089891.0
63,The Dark Knight Rises,2012,UA,164 min,"Action, Adventure",8.4,78.0,Christopher Nolan,Christian Bale,1516346,448139099.0
69,Memento,2000,UA,113 min,"Mystery, Thriller",8.4,80.0,Christopher Nolan,Guy Pearce,1125712,25544867.0
155,Batman Begins,2005,UA,140 min,"Action, Adventure",8.2,70.0,Christopher Nolan,Christian Bale,1308302,206852432.0
573,Dunkirk,2017,UA,106 min,"Action, Drama, History",7.8,94.0,Christopher Nolan,Fionn Whitehead,555092,188373161.0


### **Question 8**
Which movies have 'Drama' mentioned in their genre? ('Drama' genre wali movies ko filter karein).

In [33]:
df[df['Genre'] == "Drama"]

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
0,The Shawshank Redemption,1994,A,142 min,Drama,9.3,80.0,Frank Darabont,Tim Robbins,2343110,28341469.0
9,Fight Club,1999,A,139 min,Drama,8.8,66.0,David Fincher,Brad Pitt,1854740,37030102.0
17,One Flew Over the Cuckoo's Nest,1975,A,133 min,Drama,8.7,83.0,Milos Forman,Jack Nicholson,918088,112000000.0
20,Soorarai Pottru,2020,U,153 min,Drama,8.6,NaN,Sudha Kongara,Suriya,54995,NaN
40,American History X,1998,R,119 min,Drama,8.5,62.0,Tony Kaye,Edward Norton,1034705,6719864.0
...,...,...,...,...,...,...,...,...,...,...,...
933,"Synecdoche, New York",2008,R,124 min,Drama,7.6,67.0,Charlie Kaufman,Philip Seymour Hoffman,83158,3081925.0
934,Mysterious Skin,2004,R,105 min,Drama,7.6,73.0,Gregg Araki,Brady Corbet,65939,697181.0
941,25th Hour,2002,R,135 min,Drama,7.6,68.0,Spike Lee,Edward Norton,169708,13060843.0
946,Y tu mamá también,2001,A,106 min,Drama,7.6,88.0,Alfonso Cuarón,Maribel Verdú,115827,13622333.0


### **Question 9**
How can we sort all movies by IMDB rating in descending order? (Movies ko highest se lowest rating ke mutabiq sort karein).

In [39]:
df = df.sort_values(by= "IMDB_Rating",ascending = False)
df

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
0,The Shawshank Redemption,1994,A,142 min,Drama,9.3,80.0,Frank Darabont,Tim Robbins,2343110,28341469.0
1,The Godfather,1972,A,175 min,"Crime, Drama",9.2,100.0,Francis Ford Coppola,Marlon Brando,1620367,134966411.0
2,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,84.0,Christopher Nolan,Christian Bale,2303232,534858444.0
3,The Godfather: Part II,1974,A,202 min,"Crime, Drama",9.0,90.0,Francis Ford Coppola,Al Pacino,1129952,57300000.0
4,12 Angry Men,1957,U,96 min,"Crime, Drama",9.0,96.0,Sidney Lumet,Henry Fonda,689845,4360000.0
...,...,...,...,...,...,...,...,...,...,...,...
995,Breakfast at Tiffany's,1961,A,115 min,"Comedy, Drama, Romance",7.6,76.0,Blake Edwards,Audrey Hepburn,166544,NaN
996,Giant,1956,G,201 min,"Drama, Western",7.6,84.0,George Stevens,Elizabeth Taylor,34075,NaN
997,From Here to Eternity,1953,Passed,118 min,"Drama, Romance, War",7.6,85.0,Fred Zinnemann,Burt Lancaster,43374,30500000.0
998,Lifeboat,1944,NaN,97 min,"Drama, War",7.6,78.0,Alfred Hitchcock,Tallulah Bankhead,26471,NaN


### **Question 10**
How can we sort all movies by the number of votes they received? (No_of_Votes ke mutabiq descending order mein sort karein).

In [40]:
No_of_votes = df.sort_values(by= "No_of_Votes",ascending = False)
No_of_votes

,Series_Title,Released_Year,Certificate,Runtime,Genre,IMDB_Rating,Meta_score,Director,Star1,No_of_Votes,Gross
0,The Shawshank Redemption,1994,A,142 min,Drama,9.3,80.0,Frank Darabont,Tim Robbins,2343110,28341469.0
2,The Dark Knight,2008,UA,152 min,"Action, Crime, Drama",9.0,84.0,Christopher Nolan,Christian Bale,2303232,534858444.0
8,Inception,2010,UA,148 min,"Action, Adventure, Sci-Fi",8.8,74.0,Christopher Nolan,Leonardo DiCaprio,2067042,292576195.0
9,Fight Club,1999,A,139 min,Drama,8.8,66.0,David Fincher,Brad Pitt,1854740,37030102.0
6,Pulp Fiction,1994,A,154 min,"Crime, Drama",8.9,94.0,Quentin Tarantino,John Travolta,1826188,107928762.0
...,...,...,...,...,...,...,...,...,...,...,...
570,Raazi,2018,UA,138 min,"Action, Drama, Thriller",7.8,NaN,Meghna Gulzar,Alia Bhatt,25344,NaN
718,Scarface: The Shame of the Nation,1932,PG,93 min,"Action, Crime, Drama",7.8,87.0,Howard Hawks,Richard Rosson,25312,NaN
694,La planète sauvage,1973,U,72 min,"Animation, Sci-Fi",7.8,73.0,René Laloux,Barry Bostwick,25229,193817.0
721,God's Own Country,2017,NaN,104 min,"Drama, Romance",7.7,85.0,Francis Lee,Josh O'Connor,25198,335609.0


### **Question 11**
Which movie has the longest runtime? (Sabse lambi movie ka naam aur runtime minutes mein nikalye).

In [46]:
df['Runtime']=df['Runtime'].str.replace("min","")

In [48]:
df['Runtime']=df['Runtime'].astype(int)

In [51]:
df[df['Runtime'] == df['Runtime'].max()]['Series_Title']

140    Gangs of Wasseypur
Name: Series_Title, dtype: str

### **Question 12**
Which movies have Leonardo DiCaprio listed as the first star (Star1)?

In [59]:
df[df['Star1'] == "Leonardo DiCaprio"]['Series_Title']

8                             Inception
37                         The Departed
145                      Shutter Island
147             The Wolf of Wall Street
243                 Catch Me If You Can
343                        The Revenant
361                       Blood Diamond
652                             Titanic
879    Once Upon a Time... in Hollywood
Name: Series_Title, dtype: str

### **Question 13**
What is the average IMDB rating of all movies? (Saari movies ki average rating calculate karein).

### **Question 14**
Dataset ka poora statistical summary (`describe()` function) display karein.

### **Question 15**
'Runtime' column se 'min' text ko hata kar use proper integer (numeric) data type mein convert karein.

### **Question 16**
'Gross' column mein jo missing values (NaN) hain, unko pooray column ke median value se fill (impute) karein.

### **Question 17**
Har 'Certificate' category mein total kitni movies hain? (Value counts nikalye).

### **Question 18**
'IMDB_Rating' aur 'Meta_score' ke darmiyan correlation coefficient maloom karein.

### **Question 19**
Ek naya column banayein 'Rating_100_Scale' jisme 'IMDB_Rating' ko 10 se multiply kiya gaya ho.

### **Question 20**
Year 1990 se 2000 (both inclusive) ke darmiyan release hone wali movies ki total count aur list nikaleyin.

### **Question 21**
(Visualization) IMDB_Rating ka distribution dekhne ke liye ek neat Histogram plot karein.

### **Question 22**
(Visualization) Top 10 Directors aur unki total movies ka ek attractive Bar Chart banayein.

### **Question 23**
(Visualization) IMDB_Rating aur Gross revenue ke darmiyan ek Scatter Plot banayein taake dono ka relation dikhe.

### **Question 24**
(Visualization) Genres ka count dekhne ke liye top 5 genres ka ek clean Pie Chart banayein.

### **Question 25**
(Visualization) Meta_score ka distribution aur outliers check karne ke liye ek Box Plot banayein.

### **Question 26**
(Visualization) Dataset mein 'Certificate' wise average 'Gross' kamai ka ek Bar Plot banayein.

### **Question 27**
(Visualization) Har saal (Released_Year) kitni movies release hueen, iska ek Line Plot banayein.

### **Question 28**
(Visualization) 'No_of_Votes' ke mutabiq top 5 highest voted movies ka ek Horizontal Bar Chart banayein.

### **Question 29**
Har movie ke 'Genre' column ko split karke check karein ki total unique genres pure dataset mein kitne hain.

### **Question 30**
Dataset ko poora clean karne ke baad safely ek nayi csv file 'cleaned_movies.csv' ke naam se save karein.

### **Question 31**
Kaunsi movies hain jinki 'IMDB_Rating' 8.5 se zyada hai AUR unka 'Gross' revenue 100 million (100,000,000) se zyada hai? (Dual condition filter)

### **Question 32**
Director 'Christopher Nolan' ki un movies ke naam aur release year nikalein jinka 'Runtime' 140 minutes se zyada hai.

### **Question 33**
Aise actors (Star1) ki list nikalein jinhone 'Action' ya 'Sci-Fi' genre mein kaam kiya hai AUR unki movies ki 'IMDB_Rating' kam az kam 8.0 hai.

### **Question 34**
Year 2010 se 2020 ke darmiyan release hone wali un movies ka data filter karein jinko Certificate 'UA' ya 'A' mila hai.

### **Question 35**
Dataset mein 'Meta_score' 90 se zyada aur 'IMDB_Rating' 8.5 se kam wali unique movies kaunsi hain?

### **Question 36**
(Visualization) 'Drama' aur 'Action' genres ki movies ke 'Gross' revenue ko compare karne ke liye ek Grouped Box Plot ya Violin Plot banayein.

### **Question 37**
Har Director ki un movies ka count maloom karein jinhone 'No_of_Votes' 500,000 (5 lakh) se zyada hasil kiye hain.

### **Question 38**
Un movies ke naam aur ratings batayein jinka 'Released_Year' 2000 se pehle ka hai AUR unka 'Gross' unke overall median gross se zyada hai.

### **Question 39**
Aisi movies ka data nikalein jahan 'Star1' (Main Actor) aur 'Director' dono bilkul same person hain.

### **Question 40**
(Visualization) Top 5 highest rated movies aur unke 'Meta_score' vs 'IMDB_Rating' (scaled to 100) ka ek Side-by-Side Comparison Bar Chart banayein.